### ISSUE TITLE
Better to just use Proxy state warning when using useSnapShot

### ISSUE THREAD

**Issue by tavolafourcade** [pmndrs/eslint-plugin-valtio#26]
Hello,
I'm facing an issue with the latest version of eslint-plugin-valtio (0.4.4). The problem is that I'm using useSnapShot to re render the variable from the Store.
When I replace this implementation and call the variable directly from the Store the warning get fixed but the variable doesn't render anymore...

I tried changing the declaration of arrow function for the keyword function without success (https://github.com/pmndrs/eslint-plugin-valtio/issues/22).

`  const { isEditedMInformation } = useSnapshot(memberInformationStore)
`
vs
`memberInformationStore.isEditedMInformation`

I wasn't able to find any solution to my issue. Can somebody give me a guidance about it? Thanks!

**Comment by dai-shi** [comment]
@barelyhuman Can take a look?

**Comment by barelyhuman** [comment]
Yes sir! 🫡

@tavolafourcade 

To get better context, you are using snapshots to add into the render tree correct? 
and if you use proxy in the render tree it removes the warning? 

or are you using snapshots in a hook or a function etc , I'll need a little more code snippet or repro to see what's actually causing the rule to fire

**Comment by barelyhuman** [comment]
Assuming we are on the same page, both these tests seem to be passing and have no errors 
so I'm going to need a better explanation of what you are trying to do to fix this. (check the image below)

Also, there was a bug in the middle which could cause an unwanted warning , which you can confirm by installing this 
development version as mentioned in the comment here: https://github.com/pmndrs/eslint-plugin-valtio/pull/25#issuecomment-1128381175

If that doesn't fix, please try to give a small snippet 

![Screenshot 2022-06-11 at 7 40 10 PM](https://user-images.githubusercontent.com/43572006/173191364-5f8289fb-7286-4f6a-b17f-5b0ca483528e.png)

**Comment by tavolafourcade** [comment]
Hello there! I will better explain my problem.
Previous to the update to the latest EsLint valitio version I was using useSnapShot to render the data in the following way:
`memberInformation.ts`
<img width="1033" alt="Captura de Pantalla 2022-06-13 a la(s) 08 33 25" src="https://user-images.githubusercontent.com/63257157/173365334-131f1c6d-91b8-43a6-8e83-0ce3c7adb9a7.png">

Let's take `isEditedMInformation` as example.
When I go into my component:
`import { memberInformationStore } from 'store/memberInformation'
`
`  const { isEditedMInformation } = useSnapshot(memberInformationStore)`
And then I used it in my component:
`<Button
            color='info' onClick={isEditedMInformation || isEditedPCP ? _handleOpenModal : _handleRedirect}
            variant='light'>Cancel</Button>`

The EsLint warning made me change the way I was implementing this into the following way:
`<Button
            color='info' onClick={memberInformationStore.isEditedMInformation || isEditedPCP ? _handleOpenModal : _handleRedirect}
            variant='light'>Cancel</Button>`

So I used `memberInformationStore.isEditedMInformation`directly without using `useSnapshot` 
The warning will dissapear but I also lose the functionality expected because `isEditedMInformation` will be stuck in the initial value and would not re render when needed.

Hope this explanation is more clear and you could suggest me the way to go, thanks in advance!

**Comment by barelyhuman** [comment]
Thanks, that should be enough to find the issue

**Comment by barelyhuman** [comment]
Just to confirm, 
can you check this repo and confirm that your specific case is actually giving a warning? 
https://github.com/barelyhuman/valtio-eslint-repro-26

**Comment by barelyhuman** [comment]
@tavolafourcade 
Just to confirm, 
can you check this repo and confirm that your specific case is actually giving a warning? 
https://github.com/barelyhuman/valtio-eslint-repro-26

**Comment by barelyhuman** [comment]
@tavolafourcade
Hey, is it failing in the repo as expected ?

**Comment by haveanicedavid** [comment]
Just installed the eslint plugin, and think I might be running into something similar (not sure, but didn’t want to open a new ticket). 

My store: 
![image](https://user-images.githubusercontent.com/9052511/183768002-bacaf541-3334-4625-80b6-60758cf3472c.png)

comonent consuming: 
![image](https://user-images.githubusercontent.com/9052511/183768100-1ea74873-2ad7-4dff-a261-47a7830440cb.png)

error: 
<img width="788" alt="CleanShot 2022-08-09 at 15 50 53@2x" src="https://user-images.githubusercontent.com/9052511/183768166-466886a6-f2c2-482e-b756-de61877a58a1.png">

Changing line 60 in `Navbar.tsx` to reference the proxy itself (`value={walletStore.activeChain?.id}`) fixes the error, but then the UI doesn’t update correctly. 

Am I doing something wrong here?

**Comment by barelyhuman** [comment]
I think this is related to #32 

You might wanna try the PR specific build as mentioned here: https://github.com/pmndrs/eslint-plugin-valtio/pull/33#issuecomment-1208929745

Let me know if that fixes it for you. 


Also , simplest way to use valtio

- `snapshot for reads` 
- `proxy for writes` 
except if being used in a dep tree, eg, a useCallback or useEffect where you can execute the internal callback of `useEffect` or `useCallback` as soon as the snapshot changes. 


eg:

```js
function Comp(){
  const snap = useSnapshot(prxyState)

  return <>{snap.count}</>
}
```

```js
func Comp(){
 const snap = useSnapshot(prxyState)

  useEffect(()=>{
   console.log(snap);
},[snap])
}```



but if you are writing to the snapshot, it won't work as intended and you should write to the prxyState directly

**Comment by haveanicedavid** [comment]
> except if being used in a dep tree, eg, a useCallback or useEffect where you can execute the internal callback of useEffect or useCallback as soon as the snapshot changes.

This was my current understanding - I’ve built a decently large project with Valtio before and had no issue using in ways like this. Just installed the plugin and found the error confusing.

Looks like removing the export from the function declaration fixed it. IE
```tax
export const MyFunc = () => ()
```
throws the error, but 
```tsx
const MyFunc = () => ()
export { MyFunc }
```

Doesn’t. 

if I change from an arrow function to a named function:
```tsx
export function MyFunc() {}
```
it also fixes the error.

Thanks for the quick response! Glad to know i’m not using valtio incorrectly 😄

**Comment by barelyhuman** [comment]
yeah, so the next release should handle this, till then you can use the build from the PR if it’s blocking any CI/ git hooks. 

A previous change we made to make sure the snapshot isn’t being written to inside hooks is what’s caused this issue.
